File: mcp-code-intelligence-kotlin/bin/main/com/codeintel/memory/ingest/IngestGraphLinker.kt
/** Creates graph edges between ingested chunks from the same document. */
package com.codeintel.memory.ingest

import com.codeintel.log
import com.codeintel.memory.graph.KnowledgeGraph
import com.codeintel.memory.models.GraphEdge

/** Relation types for ingest-created edges. */
object IngestRelations {
    const val SIBLING = "SIBLING"
    const val DERIVED_FROM = "DERIVED_FROM"
}

/** Links ingested entries with graph edges (sibling + source relationships). */
class IngestGraphLinker(private val graph: KnowledgeGraph) {

    /** Create edges for a batch of chunk IDs from the same document. */
    fun linkChunks(chunkIds: List<Long>, source: String) {
        if (chunkIds.size < 2) return
        createSiblingEdges(chunkIds)
        log("IngestGraphLinker: created ${chunkIds.size - 1} edges for ${chunkIds.size} chunks from $source")
    }

    /** Create edges between a single entry and an existing source entry. */
    fun linkToSource(entryId: Long, sourceEntryId: Long) {
        val edge = GraphEdge(
            sourceId = entryId,
            targetId = sourceEntryId,
            relation = IngestRelations.DERIVED_FROM,
            weight = 0.8,
            metadata = """{"auto":"ingest"}"""
        )
        graph.addEdge(edge)
    }

    /** Calculate how many edges would be created for a chunk list. */
    fun edgeCount(chunkIds: List<Long>): Int {
        return if (chunkIds.size < 2) 0 else chunkIds.size - 1
    }

    private fun createSiblingEdges(chunkIds: List<Long>) {
        for (i in 0 until chunkIds.size - 1) {
            val edge = GraphEdge(
                sourceId = chunkIds[i],
                targetId = chunkIds[i + 1],
                relation = IngestRelations.SIBLING,
                weight = 1.0,
                metadata = """{"auto":"ingest","order":$i}"""
            )
            graph.addEdge(edge)
        }
    }
}

File: mcp-code-intelligence-nodejs/src/memory/ingest-graph-linker.ts
/**
 * Creates graph edges between ingested chunks from the same document.
 */

import { KnowledgeGraph } from './knowledge-graph.js';

/** Relation types for ingest-created edges. */
export const RELATION_SIBLING = 'SIBLING';
export const RELATION_DERIVED_FROM = 'DERIVED_FROM';

/** Links ingested entries with graph edges (sibling + source relationships). */
export class IngestGraphLinker {
  private readonly graph: KnowledgeGraph;

  constructor(graph: KnowledgeGraph) {
    this.graph = graph;
  }

  /** Create sibling edges for a batch of chunk IDs from the same document. */
  linkChunks(chunkIds: number[], source: string): void {
    if (chunkIds.length < 2) return;
    this.createSiblingEdges(chunkIds);
    console.error(`[ingest-linker] created ${chunkIds.length - 1} edges for ${chunkIds.length} chunks from ${source}`);
  }

  /** Create DERIVED_FROM edge between entry and its source. */
  linkToSource(entryId: number, sourceEntryId: number): void {
    this.graph.addEdge({
      source_id: entryId,
      target_id: sourceEntryId,
      relation: RELATION_DERIVED_FROM,
      weight: 0.8,
      metadata: '{"auto":"ingest"}',
    });
  }

  /** Calculate how many edges would be created for a chunk list. */
  edgeCount(chunkIds: number[]): number {
    return chunkIds.length < 2 ? 0 : chunkIds.length - 1;
  }

  private createSiblingEdges(chunkIds: number[]): void {
    for (let i = 0; i < chunkIds.length - 1; i++) {
      this.graph.addEdge({
        source_id: chunkIds[i],
        target_id: chunkIds[i + 1],
        relation: RELATION_SIBLING,
        weight: 1.0,
        metadata: `{"auto":"ingest","order":${i}}`,
      });
    }
  }
}
